### ISSUE TITLE
TypeScript files are published and not transpiled to JavaScript before

### ISSUE THREAD

**Issue by jelhan** [ember-bootstrap/ember-bootstrap#2106]
I noticed that we are publishing the TypeScript files. They are not transpiled to JavaScript before publishing. See `addon/components/bs-button.ts` as an example. I think that's not what we intended and may break consumers not using TypeScript.

**Comment by SanderKnauff** [comment]
For V1 addons, the addon controls it's own build behavior.
JS apps will defer the transpilation of the addon Typescript to addon itself.

I've just tested this by creating a new JS Ember project and adding ember-bootstrap to it. Adding a bs-button just works as expected.

For a live demo, check out the component docs for BsButton. https://www.ember-bootstrap.com/components/button
Our docs app does not use Typescript yet, but the pages for the converted components still work.

**Comment by jelhan** [comment]
> For V1 addons, the addon controls it's own build behavior.
> JS apps will defer the transpilation of the addon's Typescript to addon itself.

Yes. But we don't have a dependency on ember-cli-typescript or typescript or any other package, which may handle the transpilation. That's why I'm surprised that it's working.

Even if we can delay transpilation to JS to build-time of consuming app with v1 addon, I don't think we should do so. It increases technical complexity and has a negative affect on build-time performance.

**Comment by SanderKnauff** [comment]
> > For V1 addons, the addon controls it's own build behavior.
> > JS apps will defer the transpilation of the addon's Typescript to addon itself.
> 
> Yes. But we don't have a dependency on ember-cli-typescript or typescript or any other package, which may handle the transpilation. That's why I'm surprised that it's working.
> 
> Even if we can delay transpilation to JS to build-time of consuming app with v1 addon, I don't think we should do so. It increases technical complexity and has a negative affect on build-time performance.

I think it's just using `ember-cli-babel` for that. I do see that we don't pull in `@babel/plugin-typescript-transform` through the dependencies, but that package is so widely used in Ember related packages, especially the core ones, that I think most apps have it available through those.

> Even if we can delay transpilation to JS to build-time of consuming app with v1 addon, I don't think we should do so. It increases technical complexity and has a negative affect on build-time performance.

That is a valid concern. I do think that most V1 Typescript addons keep the TS in place though. Should we try to precompile the js files and remove the Typescript files from the build bundle?

I really should dive back into the V2 addon stuff. For that one atleast I'm sure we should only ship compiled js (except for the types, ofcourse).

**Comment by jelhan** [comment]
> Should we try to precompile the js files and remove the Typescript files from the build bundle?

I was surprised to see that this is not the standard. If it's not too much work, we should do it in my opinion. Maybe it's just about running `tsc` in `prepack` step and adding `*.ts` to `.npmignore`?

**Comment by jelhan** [comment]
Closing as this seems to be best practice for v1 addons.
